spv/sync.go

```diff
@@ -0,0 +1,1433 @@
+	const opf = "spv.handleBlockInvs(%v)"
+
+	blocks, err := rp.Blocks(ctx, hashes)
+	if err != nil {
+		op := errors.Opf(opf, rp)
+		return errors.E(op, err)
+	}
+	headers := make([]*wire.BlockHeader, len(blocks))
+	bmap := make(map[chainhash.Hash]*wire.MsgBlock)
+	for i, block := range blocks {
+		bmap[block.BlockHash()] = block
+		h := block.Header
+		headers[i] = &h
+	}
+
+	return s.handleBlockAnnouncements(ctx, rp, headers, bmap)
+}
+
+func (s *Syncer) lowestRescanPoint(ctx context.Context) (*chainhash.Hash, error) {
+	var rescanChainHash *chainhash.Hash
+	var rescanBlockHeight int32 = -1
+	for _, w := range s.wallets {
+		rescanPoint, err := w.RescanPoint(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if rescanPoint == nil {
+			continue
+		}
+
+		header, err := w.BlockInfo(ctx, wallet.NewBlockIdentifierFromHash(rescanPoint))
+		if err != nil {
+			return nil, err
+		}
+
+		if header.Height < rescanBlockHeight || rescanBlockHeight == -1 {
+			rescanChainHash = rescanPoint
+			rescanBlockHeight = header.Height
+		}
+	}
+
+	return rescanChainHash, nil
+}
+
+// handleTxInvs responds to the inv message created by rp by fetching
+// all unseen transactions announced by the peer.  Any transactions
+// that are relevant to the wallet are saved as unconfirmed
+// transactions.  Transaction invs are ignored when a rescan is
+// necessary or ongoing.
+func (s *Syncer) handleTxInvs(ctx context.Context, rp *p2p.RemotePeer, hashes []*chainhash.Hash) {
+	const opf = "spv.handleTxInvs(%v)"
+
+	rpt, err := s.lowestRescanPoint(ctx)
+	if err != nil {
+		op := errors.Opf(opf, rp.RemoteAddr())
+		log.Warn(errors.E(op, err))
+		return
+	}
+	if rpt != nil {
+		return
+	}
+
+	// Ignore already-processed transactions
+	unseen := hashes[:0]
+	for _, h := range hashes {
+		if !s.seenTxs.Contains(*h) {
+			unseen = append(unseen, h)
+		}
+	}
+	if len(unseen) == 0 {
+		return
+	}
+
+	txs, err := rp.Transactions(ctx, unseen)
+	if errors.Is(err, errors.NotExist) {
+		err = nil
+		// Remove notfound txs.
+		prevTxs, prevUnseen := txs, unseen
+		txs, unseen = txs[:0], unseen[:0]
+		for i, tx := range prevTxs {
+			if tx != nil {
+				txs = append(txs, tx)
+				unseen = append(unseen, prevUnseen[i])
+			}
+		}
+	}
+	if err != nil {
+		if ctx.Err() == nil {
+			op := errors.Opf(opf, rp.RemoteAddr())
+			err := errors.E(op, err)
+			log.Warn(err)
+		}
+		return
+	}
+
+	// Mark transactions as processed so they are not queried from other nodes
+	// who announce them in the future.
+	for _, h := range unseen {
+		s.seenTxs.Add(*h)
+	}
+
+	// Save any relevant transaction.
+	for walletID, w := range s.wallets {
+		relevant := s.filterRelevant(txs, walletID)
+		for _, tx := range relevant {
+			err := w.AcceptMempoolTx(ctx, tx)
+			if err != nil {
+				op := errors.Opf(opf, rp.RemoteAddr())
+				log.Warn(errors.E(op, err))
+			}
+		}
+
+		if len(relevant) > 0 {
+			s.mempoolTxs(walletID, relevant)
```

Comment:
I noticed that spv/v3.0.0 does not check the length of relevant txs before triggering this callback.
https://github.com/decred/dcrwallet/blob/dcb2562c659c09e2ddbb9913da1a1e10ab25a5df/spv/sync.go#L720-L728
```
	relevant := s.filterRelevant(txs)
	for _, tx := range relevant {
		err := s.wallet.AcceptMempoolTx(ctx, tx)
		if err != nil {
			op := errors.Opf(opf, rp.RemoteAddr())
			log.Warn(errors.E(op, err))
		}
	}
	s.mempoolTxs(relevant)
```